NEW : Add status shipped in expedition

dolibarr95 opened the pull request:
# New
Add a new status : SHIPPED

# Description
- Draft : prepare the expedition (choose the product to send from a propal or order)
- Validate : The expedition is in preparation (pick the products in warehouse, parcel, weight...) 
- Processed: The expedition is ready (to be send, chose the best carrier, tracking number...)
- Shipped: The expedition is shipped (the carrier pick the parcel)

 # Issue
#8838 

Hope everybody need this change #

dolibarr95 committed:
Add new status

Add a new status : SENDED

mwelters commented:
I think the most often used word for this status would be "SHIPPED"

dolibarr95 committed:
Add entry menu

Add new expedition status STATUS_SENDED in expedition menu

dolibarr95 committed:
Update expedition.class.php

Sended to Shipped

dolibarr95 committed:
Update eldy.lib.php

dolibarr95 commented:
Hi
yes @mwelters you're right, it's done.

dolibarr95 committed:
Add ship()

function ship() to change status 2 to status 3

dolibarr95 committed:
Add buton

Add buton to set expedition as shipped

dolibarr95 committed:
update LibStatut

I've choosed the statut9.png(grey) icon to illustrate the status SHIPPED.
Not sure if it's ok for every body.

dolibarr95 commented:
Hi
If someone can help with transiflex (i really don't understand how to do).

# EN
- StatusSendingSendedShort : Shipped
- StatusSendingSended : Shipped
# FR
- StatusSendingSendedShort : Expédiée
- StatusSendingSended : Expédiée

frederic34 commented on htdocs/expedition/class/expedition.class.php:
```diff
@@ -1590,30 +1643,35 @@ function LibStatut($statut,$mode)
 			if ($statut==0) return $langs->trans($this->statutshorts[$statut]);
 			if ($statut==1) return $langs->trans($this->statutshorts[$statut]);
 			if ($statut==2) return $langs->trans($this->statutshorts[$statut]);
+			if ($statut==23) return $langs->trans($this->statutshorts[$statut]);
```
23 ?

dolibarr95 committed:
bad copy and paste

dolibarr95 commented:
Hi
tks @frederic34  it's corrected now.

frederic34 commented on htdocs/expedition/class/expedition.class.php:
```diff
@@ -569,6 +572,56 @@ function fetch($id, $ref='', $ref_ext='', $ref_int='')
+			if (! $error)
+			{
+    				// Call trigger
+    				$result=$this->call_trigger('SHIPPING_SHIPPED',$user);
```
SHIPPING_SHIPPED, perhaps it's better to use SHIPPING_MODIFY wich is standard trigger

dolibarr95 committed:
Standard trigger name

dolibarr95 commented:
@frederic34 tks for correction, It's done

dolibarr95 committed:
Prepare to record shipped and user

But first add new column in database!

dolibarr95 committed:
Add two new column

fk_user_shipped : $user->id of the user set as shipped
date_shipped : date of the set as shipped status

dolibarr95 committed:
Add two column

fk_user_shipped : $user->id of the user set as shipped
date_shipped : date of the set as shipped status

dolibarr95 committed:
Record user and date of shipped status

dolibarr95 commented:
@eldy @frederic34 can you please give help for this conflict : 
**This branch has conflicts that must be resolved**
Where should i add this sql migration ?
```sql
ALTER TABLE llx_expedition ADD fk_user_shipped INT(11) DEFAULT NULL;
ALTER TABLE llx_expedition ADD date_shipped datetime DEFAULT NULL;
```

dolibarr95 committed:
Add status 3 in select filter

frederic34 commented:
try to add LF at end of file

dolibarr95 commented:
Hi
When I edit the doc there is already a line break (532 lines). But i will add an other one to test.

dolibarr95 committed:
add line break

frederic34 commented:
do you know how to merge upstream/develop into your branch ? Doing that allow you to resolve conflict on your machine

dolibarr95 commented:
Hi
...sorry no i really don't know

frederic34 commented:
send me an email, i will explain

dolibarr95 commented:
tks a lot it's done.
I can give you write acces to my fork if you want.

dolibarr95 committed:
Merge branch 'develop' into patch-45

dolibarr95 commented:
@frederic34  Tks a lot for your precious help.

dolibarr95 commented:
Hi
@eldy can you please check this PR please?
Tks

dolibarr95 commented:
@eldy Can you please help?
I don't understand why there is a conflict, i've resolved it but it's come back...
Tks for your help.

dolibarr95 committed:
Merge branch 'develop' into patch-45

dolibarr95 commented:
Hi ! @eldy
Hope everybody enjoy holidays :)
Can an admin help me, tag or give opinion about this PR please ?
Have a nice day

Tks

dolibarr95 committed:
Merge branch 'develop' into patch-45

dolibarr95 commented:
Conflicts solved 😀

dolibarr95 commented:
Hi
@atm-maxime can you please help ?

ksar-ksar commented:
Hi,

When this PR will be merged ?
It is a must have feature I think.

dolibarr95 commented:
Hi,
Tks for your feed back, i wait for an admin opinion.
Hope soon

ksar-ksar commented:
https://www.dolibarr.fr/forum/3-installation/62876-workflow-expedition

dolibarr95 commented:
@ksar-ksar up for this issue !
:)

dolibarr95 commented:
Hi everybody!

Little up for this Pr...
Is there an admin who can check or give advice/opinion/label/etc for this modifications ?

Have a nice monday

dolibarr95 commented:
Hi!
@FHenry can you please help ?

Have a nice day ;)

ghost commented:
Seems to be nice. My guess is why for this status is need on expedition as it exists on order

dolibarr95 commented:
Hi!
_Tks for reading and for your answer._

Because we need to know the status of an expedition only (Draft, Validated, Done, Processed) because for an order we can send partial order and also the store keeper want's to make the difference between done and proced.
More informations, examples and explanations here #8838 
tks

eldy commented:
A similar feature seems suggested to in #9660

dolibarr95 commented:
Hi
Tks for answer.
It's not the same #9660 is just about turn back in "draft" mode. What I suggest in this PR is to add a new status (4 status instead of 3 actually).
Tks for reading.

eldy commented:
To analyze also with #9660 and #10174 and #8838

eldy commented:
There is several PR to enhance this and it is not clear what will be the behaviour after (except that all of them suggest a different way of working). So to help to choose which one to retain, i updated wiki doc to describe current worklow:
https://wiki.dolibarr.org/index.php/Module_Shipments#Usage

Can you cut and paste this workflow text graph and update it to describe what will be workflow after the change (keeping the X and the number of status in parenthesis)

eldy commented:
I think we should enhance the process with something like that:
Draft ---X---> Validated ---> Closed (X = stock change, Can edit qty and lot in Draft step)
must become
Draft ----> Validated ---X---> Shipped (X = stock change, Can edit qty and lot in status Draft + Validated)

fbosman commented:
I would like to add a status: ready for shipment; the stock has been changed, but the products are still in the house.


I think we should enhance the process with something like that:
Draft ---X---> Validated ---> Closed (X = stock change, Can edit qty and lot in Draft step)
must become
Draft ----> Validated ---X---> ready for shipment ---> Shipped (X = stock change, Can edit qty and lot in status Draft + Validated)

eldy commented:
It's strange to have X (the stock change) done before a status called "ready for shipment". ready for shipment means it is somewhere in a warehouse (may be a transitional warehouse) but not yet shipped. And a stock change reflect the fact that a product is no more in the hand of company (responsability of owning the product has been transferred to someone else).  But if it is not yet shipped, it is still somewhere. Imagine you received orders to order all you stock. When all your stocks will be "ready", if it is removed from physical stock ,the virtual stock will be (0 - n) (n all open orders), so the virtual stock shows we miss a lot of product and that we must replenish. But's its wrong. Virtual stock should be 0 since we have all quantity n in stock for n ordered so we have enough to send all orders done with no need of replenishment. So it will make very difficult to work this way.

Other software solve the need with 2 different solution (depending on software):
A) Some call this the status "Reserved" (it is still the property of the company, still in stock but not yet inside a transporter truck) but not available for others. And then they report 2 types of qty when getting availability of product (total for all or total for non reserved only). Often a gaz engine, because when doing an inventory, there is no way to know what are reserved or not.
B) Some other software or processes use a different solution by adding a type "common" or "transit" for warehouse. And they ask the name of the transit warehouse (with a possible default value if it is always the same "waiting area") when changing the status to something similar to "Ready for shipment", and they record a stock movement from common warehouses to transit warehouses. Then when they report stock available, they count only if warehouse is an available warehouse and exclude transitional warehouses.

It means, enhance the process with something like that:
Draft ---X---> Validated ---> Closed (X = stock change, Can edit qty and lot in Draft step)
must become
Draft ----> Validated ---Y---> ready for shipment -- X ---> Closed/Shipped (X = stock change, Y = stock move to a transit warehouse, We can edit qty and lot in status Draft + Validated).  
Adding this "step ---Y---> ready for shipment" could become easily an option (with or without the Y, we can decide later) is for me a second step. To make evolution of this second step easier to do, we must first move the X after the validated step, to get:
Draft ----> Validated ---X----> Closed/shipped (X = stock change, Can edit qty and lot in Draft+Validated step)

It will be easier after to insert a new status.

So, we must first have a PR that move the X after before think to insert a new step.

BebZ commented:
Ahaha, Eldy we were reading the same subjects at the same time ! 
So much debate arround this subject ! 
Your POV is well described here, but I think everyone wants to modify a shipment BEFORE it is shipped (internal errors, customer wishes between the picking moment and the leave of the goods from the warehouse).

If there is something wrong after a shipment was done (so stock movement was recorded), and we need to send more, we should not go back to draft and resent the shipment. This has a lot of side effect.

Another solution to adresse that is to add option "Decrease real stocks on shipping classification to Closed / Shipped". When this option is on, you must be able to enter the batches before going to closed/shipped ?
Make it possible to modify the number shipped and lot when the shipment is in "Validated" status.

we must forecast a lot of use cases, but it make too much debates ! 
On the supplier reception side, it was less debated 
and it is much easier to make a reception and being able to put it back to draft, with stock reverts, correction, etc...

eldy commented:
Current workflow of shipment is this one:
Draft --X--> Validated ---Y---> Shipped (We can choose to have stock change into X or Y step)

The Y has been implemented in v17 or before. You can now choose X or Y.

With v18, the default mode is set to Y instead of X.

What we need now is to be able edit qty and lot in status Draft + Validated so i close this PR (it seems some are already open for that so i close this PR).
